enumerable_blocks.rb

```diff
@@ -0,0 +1,157 @@
+module Enumerable
+  # my_each
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    element = if instance_of?(Array)
+                self
+              else
+                to_a
+              end
+
+    k = 0
+    while k < element.length
+      yield element[k]
+      k += 1
+    end
+    self
+  end
+
+  # my_each_with_index
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    element = to_a
+    k = 0
+    while k < element.length
+      yield element[k], k
+      k += 1
+    end
+    self
+  end
+
+  # my_select
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    arr = []
+    my_each { |i| arr << i if yield i }
+    arr
+  end
+
+  # my_all?
+  def my_all?(argument = nil)
+    if block_given? && !argument.nil?
+      return false unless my_all?(argument) && my_all?
+    end
+    if block_given?
+      my_each { |item| return false if yield(item) }
+    elsif !block_given? && argument.nil?
+      my_each { |_item| return true unless argument == false || argument.nil? }
+    elsif !block_given? && argument.nil?
+      my_each { |item| return false if item == false || item.nil? }
+    elsif argument.is_a?(Class)
+      my_each { |_item| return false unless argument.is_a?(Class) }
+    elsif argument == Regexp
+      my_each { |item| return true if item.match(Regexp) }
+    elsif argument == Regexp
+      my_each { |item| return false unless item.match(Regexp) }
+    else
+      my_each { |item| return false if item != argument }
+    end
+    true
+  end
```

Comment:
# my_all? should:
1. returns true if the block never returns false or nil
2. returns false if the block returns false or nil
3. when no block or argument is given returns false when one of the collection members are false or nil
4. when a class is passed as an argument returns false if any of the collection is not a member of such class
5. when a Regex is passed as an argument returns true if all of the collection matches the Regex